feat(compiler): add support for shorthand property declarations in templates

crisbeto opened the pull request:
Adds support for shorthand property declarations inside Angular templates. E.g. writing `{foo, bar}` instead of `{foo: foo, bar: bar}`.

Fixes #10277.

jessicajaniuk reviewed (approved):
LGTM :cookie:

JoostK commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -945,8 +945,19 @@ export class _ParseAST {
+        } else if (this.consumeOptionalCharacter(chars.$COLON)) {
+          values.push(this.parsePipe());
+        } else {
+          const span = this.span(start);
+          const sourceSpan = this.sourceSpan(start);
```
Using `start` looks incorrect here, as that's the index of the opening `{`. We have some AST parse span tests that can verify this is correct.

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_template/GOLDEN_PARTIAL.js:
```diff
@@ -870,3 +870,54 @@ export declare class MyModule {
+    _handleClick(_value) { }
+}
+MyComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "0.0.0-PLACEHOLDER", ngImport: i0, type: MyComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
+MyComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "0.0.0-PLACEHOLDER", type: MyComponent, selector: "ng-component", ngImport: i0, template: `
```
Just a note that this needs the same `minVersion` treatment as the `?.[]` syntax.

crisbeto replied:
Just to capture what we decided in the other PR (https://github.com/angular/angular/pull/41911): we'll merge in the new syntax without bumping the `minVersion` and land a compiler change later on to deal with it in a better way.

JoostK reviewed (changes requested):
The change generally looks good, with a comment on parse spans.

I think this syntax requires special attention in the language service when renaming a property, as the shorthand syntax may need to be expanded in that case. @atscott WDYT about this case?

crisbeto commented:
I've addressed the span issue @JoostK.

JoostK commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -945,8 +945,21 @@ export class _ParseAST {
+        } else if (this.consumeOptionalCharacter(chars.$COLON)) {
+          values.push(this.parsePipe());
+        } else {
+          const spanStart = this.inputIndex;
+          const spanEnd = spanStart + key.length;
+          const span = this.span(spanStart, spanEnd);
+          const sourceSpan = this.sourceSpan(spanStart, spanEnd);
```
At this point in the parse the `key` has already been parsed, so the cursor and thus `this.inputIndex` is beyond the key. I think we'd have to capture `this.inputIndex` before parsing `key` and the `spanEnd` can be dropped, as `this.span` and `this.sourceSpan` select the proper end offset.

crisbeto commented:
Addressed the latest set of feedback and fixed the test failure.

atscott reviewed (changes requested):
I think we should add more tests around each part of the system that supports the language service:

The "template target" code ensures that we can map a position in a template to a target node in the TCB. This code is tested in `packages/language-service/ivy/test/legacy/template_target_spec.ts` (note that these tests do not pass so there needs to be an update to the template target code):

```

  describe('object literal shorthand', () => {
    fit('with one item', () => {
      debugger;
      const {errors, nodes, position} = parse(`{{ {va¦l1} }}`);
      expect(errors).toBe(null);
      const {context} = getTargetAtPosition(nodes, position)!;
      expect(context.kind).toBe(TargetNodeKind.RawExpression);
      // Maybe this should be multinode target so we target both the key and the value of the object
      // literal map
      const {node} = context as SingleNodeTarget;
      expect(node).toBeInstanceOf(e.PropertyRead);
      expect((node as e.PropertyRead).name).toBe('val1');
    });

    fit('with two shorthands', () => {
      const {errors, nodes, position} = parse(`{{ {val1, va¦l2} }}`);
      expect(errors).toBe(null);
      const {context} = getTargetAtPosition(nodes, position)!;
      expect(context.kind).toBe(TargetNodeKind.RawExpression);
      const {node} = context as SingleNodeTarget;
      expect(node).toBeInstanceOf(e.PropertyRead);
      expect((node as e.PropertyRead).name).toBe('val2');
    });

    fit('with shorthand and regular syntax', () => {
      const {errors, nodes, position} = parse(`{{ {val1: 'val1', va¦l2} }}`);
      expect(errors).toBe(null);
      const {context} = getTargetAtPosition(nodes, position)!;
      expect(context.kind).toBe(TargetNodeKind.RawExpression);
      const {node} = context as SingleNodeTarget;
      expect(node).toBeInstanceOf(e.PropertyRead);
      expect((node as e.PropertyRead).name).toBe('val2');
    });
  });
```
This brings up an immediate question:  Should we just target the property read node via a `SingleNodeTarget`? The way the TCB is currently generated, I think that's all we can do. 

We should add some tests around the TCB generation in `packages/compiler-cli/src/ngtsc/typecheck/test/span_comments_spec.ts`. This is the second part of the question above: What do we do about the divergent representations in the type check block vs what's in the template? The template code is written as `{val1}` and then generated as `{"val1": ((ctx).val1 /*span*/) /*span*/}` in the TCB. The language service leverages the TCB for most of the actions we perform. Because the TCB output is different than the representation in the template, it would be hard to get everything fully correct (renames, go to definition, etc.). This might just be something we have to be okay with and just do actions off of the property read. We might need to add a new template target type that can detect this case and not allow renaming. Or we just allow a rename from the property read and be okay that we get it wrong where template shorthand is used (which hopefully isn't many places and at least the compiler will detect them and complain).

After the template node is found, the language service builds a symbol for the template node. This part is tested in `packages/compiler-cli/src/ngtsc/typecheck/test/type_checker__get_symbol_of_template_node_spec.ts`. Please add a test there to ensure we build the correct symbol for this case.

Lastly, we have specific functions in the language service for LSP operations (quick info, definitions, etc) using the template symbol information. Generally if the symbol was built correctly, the other parts _should_ also work. Here's a test for quick info (`packages/language-service/ivy/test/quick_info_spec.ts`):

```

  // TODO(atscott): This should be able to be added to the previous section as long as the project
  // has a different name than the one in the beforeEach but something goes wrong in the test
  // environment
  describe('non-skeleton project', () => {
    beforeEach(() => {
      initMockFileSystem('Native');
      env = LanguageServiceTestEnv.setup();
    });
    fit('should work for shorthand object literal', () => {
      project = env.addProject(
          'test', {
            'app.ts': `
            import {Component, NgModule} from '@angular/core';
            import {CommonModule} from '@angular/common';

            @Component({
              selector: 'some-cmp',
              templateUrl: './app.html',
            })
            export class SomeCmp {
              val1 = 'one';
              val2 = 2;

              doSomething(obj: {val1: string, val2: number}) {

              }
            }

            @NgModule({
              declarations: [SomeCmp],
              imports: [CommonModule],
            })
            export class AppModule{
            }
       `,
            'app.html': `{{doSomething({val1, val2})}}`,
          },
          {strictTemplates: true});
      env.expectNoSourceDiagnostics();
      project.expectNoSourceDiagnostics();

      const template = project.openFile('app.html');
      template.moveCursorToText('val¦1');
      const quickInfo = template.getQuickInfoAtPosition();
      expect(toText(quickInfo!.displayParts)).toEqual('(property) val1: string');
      template.moveCursorToText('val¦2');
      const quickInfo2 = template.getQuickInfoAtPosition();
      expect(toText(quickInfo!.displayParts)).toEqual('(property) val2: number');
    });
  });
```
You propably want to add a similar test to `definitions_spec` in the same folder to ensure "go to definition" works properly as well as the `references_and_rename_spec`.

crisbeto commented:
@atscott all of the feedback should be covered in the latest fixup commit. I've also included more tests for safe keyed reads since I didn't get the chance to add them before #41911 got merged in.

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/test/type_checker__get_symbol_of_template_node_spec.ts:
```diff
@@ -682,6 +702,14 @@ runInEachFileSystem(() => {
         });
+
+        it('literal map with shorthand property', () => {
+          const literalMap = interpolation.expressions[2] as LiteralMap;
```
Could you update this or add a test that gets the symbol for the property in the map rather than the map itself? I don't know if, in practice, the vscode extension requests symbols for the whole map.

atscott commented on packages/language-service/ivy/test/legacy/definitions_spec.ts:
```diff
@@ -458,6 +458,18 @@ describe('definitions', () => {
+
+      const [def] = definitions;
+      expect(def.textSpan).toEqual('name');
+      expect(def.contextSpan).toContain('name: string;');
```
Can you add an assert for `def.fileName` as well? This part is currently not working - the "go to definition" attempts to go to the typecheck file which doesn't exist on disk.

crisbeto commented:
The latest set of feedback has been addressed.

JoostK reviewed (approved):
LGTM for fw-compiler

atscott commented on packages/language-service/ivy/test/legacy/definitions_spec.ts:
```diff
@@ -458,6 +458,19 @@ describe('definitions', () => {
+      expect(def.fileName).toContain('/app/app.component.ts');
+      expect(def.contextSpan).toContain('name: string;');
+    });
   });
```
Ah, interesting. The above test _does_ work for go to definition. I was actually testing this case locally: 
`{{ {name} }}`.

Could you add another test for that case? I think the `fileName` for that one is going to be the typecheck file which will give this error if you try to go to definition:
![image](https://user-images.githubusercontent.com/479713/121227560-8c2aa800-c840-11eb-9be8-af687fe548e6.png)

It seems the symbol you're going to is the key of the object rather than the property read. This'll happen because they have the same source span `"name": ((ctx).name /*49,58*/) /*49,58*/` and we just target the first one we find.

We should update the code in `template_symbol_builder` to be 

```
function getSymbolOfTemplateExpression...
   ...
    const anyNodeFilter = (n: ts.Node): n is ts.Node => true;
    const filter =
        expression instanceof PropertyRead ? ts.isPropertyAccessExpression : anyNodeFilter;
    let node = findFirstMatchingNode(this.typeCheckBlock, {withSpan, filter});
    ...
```

This will also update the test you have for that part to be 
`expect(program.getTypeChecker().symbolToString(symbol.tsSymbol!)).toEqual('foo');` 
instead of
`expect(program.getTypeChecker().symbolToString(symbol.tsSymbol!)).toEqual('__type');`
which I think is what we want.

crisbeto commented:
The last comment has been addressed @atscott.

atscott reviewed (approved):
👍  LGTM

atscott commented:
[presubmit](http://test/OCL:378894848:BASE:378894881:1623431003756:21964cd6)

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/test/span_comments_spec.ts:
```diff
@@ -129,6 +138,13 @@ describe('type check blocks diagnostics', () => {
               '(null as any ? (((ctx).a /*3,4*/) /*3,4*/)!.method /*6,12*/(((ctx).b /*13,14*/) /*13,14*/) : undefined) /*3,15*/');
     });
 
+    it('should annotate safe keyed reads', () => {
```
This test appears unrelated to everything else in this PR. Is it here because prior designs broke this feature?

Either way, I would recommend extracting the addition of this test as a separate commit prior to the main commit, to avoid confusion.

crisbeto replied:
These are tests that I missed in my safe keyed reads PR (https://github.com/angular/angular/pull/41911). I decided to add them here since this PR touches the same test files.

alxhub replied:
I think we should squash the commits and then extract these tests into a separate commit, honestly.

It's not a big deal, but I think in the name of having a clean history, it makes sense not to bundle unrelated tests.

To avoid another round trip and blocking this PR, I'll go ahead and do that for you since I think it's ready to merge otherwise.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts:
```diff
@@ -482,8 +482,13 @@ export class SymbolBuilder {
 
-    let node = findFirstMatchingNode(
-        this.typeCheckBlock, {withSpan, filter: (n: ts.Node): n is ts.Node => true});
+    let node =
```
This operation took me a while to unravel. Two comments:

1) Why is it now necessary to prioritize `ts.PropertyAccessExpression` nodes over other nodes in the case of `PropertyRead` operations? This seems like it might be an avoidable change. If not, I think we should document here why `PropertyRead` is special-cased.

2) We should also restructure this flow to be much more obvious - the `||` operation I think is a bit too hard to follow. Extracting the node search into a utility function could be extremely helpful here. Something like:

```typescript
function getExpressionTsNode(...): ts.Node|null {
  if (expression instanceof PropertyRead) {
    // Prioritize looking for a `ts.PropertyAccessExpression` for `PropertyRead` expressions, because
    // ...
    const node = findFirstMatchingNode(..., ts.isPropertyAccessExpression);
    if (node !== undefined) {
      return node;
    }
  }

  // In most cases, find the first matching node of any kind.
  return findFirstMatchingNode(..., anyNodeFilter) ?? null;
}
```

alxhub commented:
If you wouldn't mind, I think we could also use a test here for a case like:
```html
<div #foo></div>
{{ m({foo}) }}
```
just to verify that a shorthand reference to a property defined in the template itself still works.

alxhub commented:
(and apologies for the last-minute change requests! this PR is 👍 otherwise, this is just a confusing part of the system and i want to make sure it's as clear as possible for people in the future)

crisbeto commented:
I've addressed the latest set of feedback @alxhub.

alxhub reviewed (approved):
(I've pushed an update that splits out the extra tests for keyed reads into a separate commit, to keep the history clean)

crisbeto committed:
test(compiler-cli): add additional safe keyed read tests

This commit adds some tests that were mistakenly omitted from the original
change for safe keyed reads/writes.

crisbeto committed:
feat(compiler): add support for shorthand property declarations in templates

Adds support for shorthand property declarations inside Angular templates. E.g. doing `{foo, bar}` instead of `{foo: foo, bar: bar}`.

Fixes #10277.

dylhunn commented:
@AndrewKushnir suggests running a TGP, I will do so today.

dylhunn commented:
This CL is on the 1pm TAP train, should have results this afternoon.

dylhunn commented:
TGP is green.
